### ISSUE TITLE
Successful Mail Delivery Report due to spoofed return-path mydomain spam

### ISSUE THREAD

**Issue by Stephano2000** [virtualmin/virtualmin-gpl#543]
I can't say this is a Virtualmin issue per se, but it's worth having a look at as it is an actual case.
Some spam emails (from external sources) with:
- Spoofed Return-Path: sender@mydomain
- Spoofed From: sender@mydomain
- To: recipient@mydomain

are softfailing the spf test and landing in the spam folder (with a very high score) as expected. However it looks like the spoofed emails are sent with a DSN request, prompting postfix to send an unwarranted DSN to sender@mydomain.
The net result is that sender@mydomain receives an annoying "Successful Mail Delivery Report" for each such spam email.
Is there anyway to prevent postfix from sending the DSN for an email that softfails spf and is classified as spam by SpamAssassin? After all it's a spoofed email from mydomain to mydomain!

**Comment by iliajie** [comment]
Hello,

Yeah, this is something related to Postfix configuration. We try to stick to system defaults as much as possible, which should work for most users. However, specific configurations can be implemented manually.

**Comment by chris001** [comment]
One way to prevent Postfix from sending a DSN email for a spam email, you can configure SpamAssassin to add the "`discard`" header to the email. To do this, add this line to `/etc/mail/spamassassin/local.cf`:

```
add_header all DSN: Discarded
```

With this configuration, when SpamAssassin classifies an email as spam, it will add the "`discard`" header to the email, which tells Postfix to discard the email and not send a DSN email.

**Comment by chris001** [comment]
If you want to deliver the spam email to junk or spam folder, then try this. Configure SpamAssassin to add a special header to the email, and then configure Postfix to not send a DSN email for emails with that header.

1. Configure SpamAssassin to add a header to emails with a high spam score. 

In the SpamAssassin configuration file (usually `/etc/mail/spamassassin/local.cf`), add the following line:

```
add_header all X-Spam-Flag YES
```

This will add a header called "X-Spam-Flag" to all emails that SpamAssassin scores as spam.

2. Configure Postfix to not send a DSN email for emails with the "X-Spam-Flag" header. 

In the Postfix configuration file (usually `/etc/postfix/main.cf`), add the following line to the `smtpd_recipient_restrictions` section:

```
smtpd_delay_reject = yes
smtpd_soft_error_limit = 10
smtpd_hard_error_limit = 20
smtpd_error_sleep_time = 1s
```

3.    Create a new file called `/etc/postfix/header_checks` with the following contents:

```
/^X-Spam-Flag: YES/ DISCARD X-Spam-Flag: Your message was classified as spam and not delivered to the recipient.
```

This tells Postfix to not send a DSN email for any emails with the "X-Spam-Flag: YES" header, and instead discard the email and send a custom message in the response.

With this configuration, when an email arrives, SpamAssassin will score it for spam, and if the score is high enough, it will add the "X-Spam-Flag" header. Postfix will then see the "X-Spam-Flag" header and discard the email without sending a DSN email, but still deliver the email to the spam or junk folder.

**Comment by Stephano2000** [comment]
> Hello,
> 
> Yeah, this is something related to Postfix configuration. We try to stick to system defaults as much as possible, which should work for most users. However, specific configurations can be implemented manually.

I'd rather always stick to using Virtualmin's UI, and not edit config files directly. Webmin/SpamAssassin's configuration has a Header Checks section, but it seems it's actually a Mime Header Checks.

**Comment by Stephano2000** [comment]
> ```
> add_header all X-Spam-Flag YES
> ```
This adds an ```X-Spam-Flag YES``` header to all mail. I guess it should be instead ```add_header spam X-Spam-Flag YES```, which is the default for SpamAssassin anyway

> ```
> /^X-Spam-Flag: YES/ DISCARD X-Spam-Flag: Your message was classified as spam and not delivered to the recipient.
> ```
Won't this discard the mail altogether?

**Comment by chris001** [comment]
Yes.  Try this instead.  In the Postfix configuration file (usually `/etc/postfix/main.cf`), add the following lines:

```
notify_classes = resource, software
bounce_notice_recipient =
2bounce_notice_recipient =
```
This should suppress the DSN email for resource and software notifications.

**Comment by Stephano2000** [comment]
It seems the sensible way is to disable DSN altogether:
`smtpd_discard_ehlo_keywords = silent-discard, dsn`

I opened an issue in webmin section:
(https://github.com/webmin/webmin/issues/1879)

**Comment by iliaross** [comment]
Reopened, as this was closed by an automated cleanup run.

⚠️ If the issue is resolved, please take a moment to close it to help keep the project organized.
